tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15360,8 +15360,30 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
 
 void DiagnoseEntry(Sema &S, FunctionDecl *FD) {
   auto Attr = FD->getAttr<HLSLShaderAttr>();
-  if (!Attr)
-    return;
+  if (!Attr) {
+    // if this is the Entry FD, then add the target profile
+    // shader attribute to the FD and carry on with validation
+    // otherwise, with no shader attribute, just return
+    const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;
+    if (EntryPointName.empty()) {
+      return;
+    }
+
+    
+    // if this FD isn't the entry point, then there's no
+    // shader attribute to work with, so just return
+    if (EntryPointName != FD->getIdentifier()->getName()) {
```

Comment:
This will need to be more specific about the scope of the decl, otherwise we could match a function by the same name outside global scope.